src/util/expr_iterator.h

```diff
@@ -0,0 +1,242 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "invariant.h"
+
+// Forward declarations - table of contents
+
+/// \file Forward depth-first search iterators
+/// These iterators' copy operations are expensive, so use auto&, and avoid
+/// std::next(), std::prev() etc.
+///
+/// Incrementable only by calling ++it (to avoid accidental copying, as
```

Comment:
http://www.cplusplus.com/reference/iterator/ Says that post-increment should work on all categories of iterator. I agree that pre-increment is usually preferable, but this may break code which expects post-increment to work.